substrate/frame/contracts/src/wasm/mod.rs

```diff
@@ -1878,32 +1885,37 @@ mod tests {
 		assert_ok!(execute(CODE_VALUE_TRANSFERRED, vec![], MockExt::default()));
 	}
 
-	const START_FN_ILLEGAL: &str = r#"
+	const START_FN_DOES_RUN: &str = r#"
 (module
-	(import "seal0" "seal_return" (func $seal_return (param i32 i32 i32)))
+	(import "seal0" "seal_deposit_event" (func $seal_deposit_event (param i32 i32 i32 i32)))
 	(import "env" "memory" (memory 1 1))
 
 	(start $start)
-	(func $start
-		(unreachable)
+	(func $start 
+		(call $seal_deposit_event
+			(i32.const 0) ;; Pointer to the start of topics buffer
+			(i32.const 0) ;; The length of the topics buffer.
+			(i32.const 0) ;; Pointer to the start of the data buffer
+			(i32.const 13) ;; Length of the buffer
+		)
 	)
 
-	(func (export "call")
-		(unreachable)
-	)
+	(func (export "call"))
 
-	(func (export "deploy")
-		(unreachable)
-	)
+	(func (export "deploy"))
 
-	(data (i32.const 8) "\01\02\03\04")
+	(data (i32.const 0) "\00\01\2A\00\00\00\00\00\00\00\E5\14\00")
 )
 "#;
 
 	#[test]
-	fn start_fn_illegal() {
-		let output = execute(START_FN_ILLEGAL, vec![], MockExt::default());
-		assert_err!(output, <Error<Test>>::CodeRejected,);
+	fn start_fn_does_run() {
+		let mut ext = MockExt::default();
+		execute(START_FN_DOES_RUN, vec![], &mut ext).unwrap();
+		assert_eq!(
+			ext.events[0].1,
+			[0x00_u8, 0x01, 0x2a, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xe5, 0x14, 0x00]
+		);
```

Comment:
start runs for both call and deploy? maybe we check that in the assert too

Reply:
It should really be independent of the called export, i.e. should work for any export:

```rust
		match instance.start(&mut store) {
			Ok(instance) => {
				let exported_func = instance
					.get_export(&store, function.identifier())
					.and_then(|export| export.into_func())
					.ok_or_else(|| {
						log::error!(target: LOG_TARGET, "failed to find entry point");
						Error::<T>::CodeRejected
					})?;

				// ...
			},
			Err(err) => process_result(store, Err(err)),
		}
```

However I agree, the same testcase for `deploy` wouldn't hurt :)

Reply:
Yes the start function always runs once after a wasm instance is created. Will add a test.

Reply:
Added the test.